pkg/infra/observer.go

```diff
@@ -5,62 +5,55 @@ import (
 	"time"
 
 	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 )
 
 type Observer struct {
 	d      peer.Deliver_DeliverFilteredClient
 	logger *log.Logger
-	signal chan error
 }
 
-func CreateObserver(channel string, node Node, crypto *Crypto, logger *log.Logger) *Observer {
+func CreateObserver(channel string, node Node, crypto *Crypto, logger *log.Logger) (*Observer, error) {
 	deliverer, err := CreateDeliverFilteredClient(node)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	seek, err := CreateSignedDeliverNewestEnv(channel, crypto)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	if err = deliverer.Send(seek); err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	// drain first response
 	if _, err = deliverer.Recv(); err != nil {
-		panic(err)
+		return nil, err
 	}
 
-	return &Observer{d: deliverer, signal: make(chan error, 10), logger: logger}
+	return &Observer{d: deliverer, logger: logger}, nil
 }
 
-func (o *Observer) Start(N int, now time.Time) {
-	defer close(o.signal)
+func (o *Observer) Start(N int, errorCh chan error, finishCh chan struct{}, now time.Time) {
+	defer close(finishCh)
 	o.logger.Debugf("start observer")
 	n := 0
 	for n < N {
 		r, err := o.d.Recv()
 		if err != nil {
-			o.signal <- err
+			errorCh <- err
```

Comment:
this alters original behavior:
old: simply logs error here and proceed
new: process exits

could you confirm this wouldn't cause any problem? (i vaguely remember that peer might throw benign errors?)

Reply:
https://github.com/guoger/stupid/blob/30c148db9f13bcdf2f2510416d9e8cc806c9dbf1/cmd/stupid/main.go#L100-L112

old, error here and processed in waiting process.
new, the error handle and process in main process and exited by errorCh

old: the main process waiting for signal channel.
new: move the signal channel to struct and using error channel as receiver of errors.
the main process waiting for singal channel or any error in errorCh

@guoger